### PR TITLE
ci: validate-migration: compile with clang and support valgrind

### PR DESCRIPTION
**Pull request by ggilmore**
This PR tweaks our CI pipeline to:

- build p4-fusion with `clang`
- use valgrind's [memcheck](https://valgrind.org/info/tools.html#helgrind) tool to detect memory manage
- use valgrind's [helcheck](https://valgrind.org/info/tools.html#helgrind) tool to detect race conditions
- compile the OpenSSL library with correct the `-fsanitizer=`  flags (as required by some sanitizers: https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual#non-instrumented-code)
-

Notes:
- In `gcc`, thread sanitizer reports an error. `Clang`'s thread sanitizer doesn't find anything though (which is disconcerting). 
- Some could say that having both [ThreadSanitizer](https://clang.llvm.org/docs/ThreadSanitizer.html) and Valgrind's Helcheck is duplicative (same for AddressSanitizer and Valgrind's Memcheck). However, most of the saniziation tools require that the entire binary + its libraries be re-compiled with `-fsanitizer=address` (etc.). We can't do this for the p4 library since we don't have access to its source code. It's possible that the sanitizers will report false positives / negatives due to this. However, Valgrind has no such restriction due to the way it works (with the downside that it's much, much slower). I think it's worth leaving both enabled for now. 
- The pipleine's logic has been re-written to use [sscache](https://github.com/mozilla/sccache) for safer/faster build caching. (Thanks @varungandhi-src !) 

## Test plan

CI

### CODE REVIEW

File: .github/workflows/run-test.sh [hunk 1]
```diff
@@ -52,7 +52,6 @@ Try using 'p4 -u ${P4USER} login -a' to generate a session ticket.
 See '${handbook_link}' for more information.
 END
 
-    exit 1
```
Comment (eseliger) [inline .github/workflows/run-test.sh:55]: intentionally removed?

**Comment by ggilmore** [comment]
Helgrind is reporting a race condition here, but I haven't looked into whether or not this one is legitimate quite yet: https://github.com/sourcegraph/p4-fusion/actions/runs/6805606733/job/18505447109?pr=11

File: .github/workflows/validate-migration.yaml [hunk 2]
```diff
@@ -57,7 +115,47 @@ jobs:
...
+            gcc \
+            g++ \
+            clang \
+            libc6-dbg \
+            libc++-dev \
+            libc++abi-dev \
+            llvm \
+            ninja-build \
+            gettext
+          sudo apt remove valgrind --yes # remove any existing valgrind installation since we're compiling it from source
```
Comment (varungandhi-src) [inline .github/workflows/validate-migration.yaml:131]: What is the rationale for compiling `valgrind` from source instead of using the pre-built version? The one in Ubuntu 22.04 is about 2 years old, but that should be fine.

File: .github/workflows/validate-migration.yaml [hunk 3]
```diff
@@ -5,49 +5,107 @@ on: [ push ]
...
+            ;;
+          
+            valgrindMemcheck)
+              echo "USE_VALGRIND=true" >> $GITHUB_ENV
+              echo "CMAKE_CXX_FLAGS=-g" >> $GITHUB_ENV
+              echo "VALGRIND_FLAGS=--tool=memcheck --leak-check=full --show-leak-kinds=all --track-origins=yes" >> $GITHUB_ENV
+            ;;
+          
+            valgrindHelgrind)
+              echo "USE_VALGRIND=true" >> $GITHUB_ENV
```
Comment (varungandhi-src) [inline .github/workflows/validate-migration.yaml]: Suggestion: Overall, I think it would be simpler to set some variables in the branches, and then unconditionally add them to the environment, rather than the current logic around conditionally adding variables to the environment.

You can also get rid of the `USE_VALGRIND` variable that way, and instead check if `VALGRIND_FLAGS` is empty or not.

File: .github/workflows/validate-migration.yaml [hunk 4]
```diff
@@ -5,49 +5,107 @@ on: [ push ]
...
-          - addressSanitizer
-          - threadSanitizer
-          #          disabled until we figure out how to build with clang in github actions
-          #          - memorySanitizer
-          - leakSanitizer
-          - undefinedSanitizer
+        tool: # Note: If you add or change the name of tools here, you must also adjust the logic in the "Prepare environment for specific tool" step
+          - addressSanitizer # https://github.com/google/sanitizers/wiki/AddressSanitizer
+          - threadSanitizer # https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual
+          - undefinedSanitizer # https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html
```
Comment (varungandhi-src) [inline .github/workflows/validate-migration.yaml]: You can combine the UBSan and ASan runs into a single one to save on a job.

File: .github/workflows/validate-migration.yaml (hunk 2 above)
Comment (varungandhi-src) [reply .github/workflows/validate-migration.yaml:131]: Sorry, found this answered below.

File: .github/workflows/validate-migration.yaml [hunk 5]
```diff
@@ -5,49 +5,107 @@ on: [ push ]
...
+          # for more information on the method to set environment variables.
+          
+          bash <<EOF
+            set -euxo pipefail
+          
+            case ${{ matrix.tool }} in
+            
+            noSanitizer)
+              echo "USE_VALGRIND=false" >> $GITHUB_ENV
+              echo "CMAKE_CXX_FLAGS=" >> $GITHUB_ENV
```
Comment (varungandhi-src) [inline .github/workflows/validate-migration.yaml]: I think you should start with `CMAKE_CXX_FLAGS="-O1 -g"` as the default, and append more things to it in the different branches as needed. `-O1` is to save time, and `-g` should provide better stack traces if there is a crash (regardless of whether it is a build with sanitizers or without).

File: .github/workflows/validate-migration-scripts/compile-openssl.sh [hunk 6]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env bash
+
+SCRIPT_ROOT="$(dirname "$(realpath "${BASH_SOURCE[0]}")")"
+cd "${SCRIPT_ROOT}/../../.."
```
Comment (varungandhi-src) [inline .github/workflows/validate-migration-scripts/compile-openssl.sh]: These two lines seem redundant both here and in the `compile-valgrind.sh` script, since you're dong `pushd` right away after it. Can you instead add a check that the `OPENSSL_SOURCE_DIR` provided is an absolute directory?

File: .github/workflows/validate-migration-scripts/compile-openssl.sh [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+read -r -a array <<<"${OPENSSL_FLAGS:-}"
+for flag in "${array[@]}"; do
+  if [[ -n "${flag}" ]]; then
+    CONFIG_ARGS+=("$flag")
+  fi
+done
+
+./config "${CONFIG_ARGS[@]}"
+make "-j$(nproc)"
```
Comment (varungandhi-src) [inline .github/workflows/validate-migration-scripts/compile-openssl.sh]: nit: Maybe `make -j "$(nproc)"` instead? It seems a little weird to see the `-j` inside the string. It's also not obvious that this syntax is supported based on `make --help` on Linux:

```
-j [N], --jobs[=N]          Allow N jobs at once; infinite jobs with no arg.
```

File: .github/workflows/validate-migration.yaml [hunk 8]
```diff
@@ -73,29 +171,43 @@ jobs:
...
-          P4CLIENT: "validate-migration-${{ github.ref_name }}-${{ github.run_number }}-${{ matrix.sanitizer }}" # the temporary Perforce client name to use while validating the git migration
+          P4CLIENT: "validate-migration-${{ github.ref_name }}-${{ github.sha }}-${{ github.run_number }}-${{ matrix.tool }}" # the temporary Perforce client name to use while validating the git migration
           OPENSSL_ROOT_DIR: ${{ env.OPENSSL_INSTALL_DIR }}
-          CMAKE_CXX_FLAGS: "${{ matrix.CMAKE_CXX_FLAGS }} -O0"
\ No newline at end of file
+          CMAKE_CXX_COMPILER: "clang++"
+          CMAKE_C_COMPILER: "clang"
+          CMAKE_C_COMPILER_LAUNCHER: "sccache"
+          CMAKE_CXX_COMPILER_LAUNCHER: "sccache"
+          CMAKE_CXX_FLAGS: "${{ env.CMAKE_CXX_FLAGS }} -stdlib=libc++ -fno-omit-frame-pointer -O0"
```
Comment (varungandhi-src) [inline .github/workflows/validate-migration.yaml]: Suggestion: Please consolidate all the arguments being set within a single step/job. E.g. where I suggested adding `-O1` earlier, you can add the `-stdlib` and `-fno-omit-frame-pointer` args there.

File: .github/workflows/validate-migration.yaml [hunk 9]
```diff
@@ -57,7 +115,47 @@ jobs:
...
+            git \
+            make \
+            cmake \
+            gcc \
+            g++ \
+            clang \
+            libc6-dbg \
+            libc++-dev \
+            libc++abi-dev \
+            llvm \
```
Comment (varungandhi-src) [inline .github/workflows/validate-migration.yaml]: Mention that `llvm` is being installed for `llvm-symbolizer` which is needed for symbolicating stack traces.

File: .github/workflows/validate-migration.yaml [hunk 10]
```diff
@@ -73,29 +171,43 @@ jobs:
...
+          .github/workflows/validate-migration-scripts/compile-openssl.sh
+        env:
+          CC: "sccache clang"
+          CXX: "sccache clang++"
 
       - name: "Run validate migration script"
         run: |
-          sudo apt-get install gettext --yes # for envsubst 
-          .github/workflows/run-test.sh
+          .github/workflows/validate-migration-scripts/run-test.sh
```
Comment (varungandhi-src) [inline .github/workflows/validate-migration.yaml]: Maybe this script and YAML step need to be renamed? The usage of "run" makes it sound like you're just going to run the built code, but this step is doing both building and running. So maybe `Build p4-fusion and test migration validation`?

File: .github/workflows/validate-migration.yaml [hunk 11]
```diff
@@ -73,29 +171,43 @@ jobs:
...
+          CMAKE_C_COMPILER_LAUNCHER: "sccache"
+          CMAKE_CXX_COMPILER_LAUNCHER: "sccache"
+          CMAKE_CXX_FLAGS: "${{ env.CMAKE_CXX_FLAGS }} -stdlib=libc++ -fno-omit-frame-pointer -O0"
+          UBSAN_OPTIONS: "print_stacktrace=1" # print stack trace on undefined behavior
+          ASAN_OPTIONS: "check_initialization_order=1" # check initialization order of globals
+
+      - name: "Ensure p4 client is deleted"
+        if: always()
+        run: |
+          p4 client -f -Fs -d "${P4CLIENT}" || true
```
Comment (varungandhi-src) [inline .github/workflows/validate-migration.yaml]: Is the client started with a timeout or something? Is this step guaranteed to run even if someone cancels a running job midway using the web UI (and if so, would that be bad?)?

File: generate_cache.sh [hunk 12]
```diff
@@ -42,5 +44,19 @@ else
   )
 fi
 
+# set the compiler launcher if it is set
+
```
Comment (varungandhi-src) [inline generate_cache.sh]:
```suggestion
```

Nit: The code explains itself

**Comment by varungandhi-src** [comment]
> In gcc, thread sanitizer reports an error. Clang's thread sanitizer doesn't find anything though (which is disconcerting).

Could you share the error? Thread Sanitizer is not guaranteed to find all data races.

**Review by varungandhi-src — APPROVED** [review]
Included some minor feedback inline.

**Review by eseliger — APPROVED** [review]

File: .github/workflows/validate-migration.yaml (hunk 5 above)
Comment (ggilmore) [reply .github/workflows/validate-migration.yaml]: I was hestiant to do this since valgrind's documentation recommends using `-O0`, but it does run twice as fast under memcheck with `-O1`. I'll leave a comment in the workflow definition saying that if `memcheck` reports an issue but it's hard to follow the stack trace - then it's worth probably temporarily shifting the level back to `-O0`.

Comment (ggilmore) [reply .github/workflows/validate-migration.yaml]: I was hestiant to do this since valgrind's documentation recommends using `-O0`, but it does run twice as fast under memcheck with `-O1`. I'll leave a comment in the workflow definition saying that if `memcheck` reports an issue but it's hard to follow the stack trace - then it's worth probably temporarily shifting the level back to `-O0`.

File: .github/workflows/validate-migration.yaml (hunk 11 above)
Comment (ggilmore) [reply .github/workflows/validate-migration.yaml]: Our perforce license allows for a limited number of [clients](https://www.perforce.com/manuals/p4guide/Content/P4Guide/configuration.workspace.html) to be defined at any given time (~20). For each job in the matrix, we creature a unique client (with a name that's unique to each job using a mix of the job number, tool we're using, sha, etc.). 

I have [logic in the run-test script](https://github.com/sourcegraph/p4-fusion/blob/94cc616797d2bf713ffbbb27e4625ecd253f87ee/.github/workflows/validate-migration-scripts/run-test.sh#L21-L42) to ensure that we delete the client whenever the script exits using a trap. However, if the job as a whole is cancelled, there is no guaranteed that trap will execute (leaving behind a client that consumes one of our licenses). This extra step that I added here will _always_ run (through the use of the `if: always()` even if the job is cancelled - ensuring that the client is cleaned up in most circumstances.

Comment (ggilmore) [reply .github/workflows/validate-migration.yaml]: >Is this step guaranteed to run even if someone cancels a running job midway using the web UI (and if so, would that be bad?)?

Yes, it is guaranteed to run according to https://docs.github.com/en/actions/learn-github-actions/expressions#always. This is the explicit intent (to ensure that we don't consume one of the clients on our license). This is not bad, since we create a new client for each commit / job.

**Comment by ggilmore** [comment]
> > In gcc, thread sanitizer reports an error. Clang's thread sanitizer doesn't find anything though (which is disconcerting).
> 
> Could you share the error? Thread Sanitizer is not guaranteed to find all data races.

https://github.com/sourcegraph/p4-fusion/actions/runs/6750817141/job/18353865452 (for commit https://github.com/sourcegraph/p4-fusion/commit/7c2ae08b4fe60c6d7e88b40881bcad025e360276)

**Comment by ggilmore** [comment]
Merging this now. We can follow up on the helgrind reports afterwards.

**Comment by varungandhi-src** [comment]
> [sourcegraph/p4-fusion/actions/runs/6750817141/job/18353865452](https://github.com/sourcegraph/p4-fusion/actions/runs/6750817141/job/18353865452) (for commit https://github.com/sourcegraph/p4-fusion/commit/7c2ae08b4fe60c6d7e88b40881bcad025e360276)

For the data race, it seems like a problem in the Perforce code itself, which we don't have access to:

```
      #0 setlocale ../../../../src/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:3353 (libtsan.so.0+0x58beb)
      #1 Enviro::GetLocale(StrBuf&, Error*) ../p4/sys/enviro.cc:1015 (p4-fusion+0x13150f)
```

We were assuming that calling the client API from multiple threads is safe, but that doesn't seem to be the case. It's bizarre though that a function called `GetLocale` would actually call `setlocale`. 🙃

Do we have contact information for someone at Perforce? This seems like a bug in their code, so we should report it.

In the meantime, I wonder if it is possible to create our own `setlocale` wrapper which takes a global lock before calling out to the actual `setlocale`. It seems like it should be possible somehow. For example, see https://github.com/ccurtsinger/interpose
